README.md
# 🕹️ xl
## [pub.dev Listing](https://pub.dev/packages/xl) | [API Doc](https://pub.dev/documentation/xl/latest) | [GitHub](https://github.com/Zabadam/xl)
#### API References: [`XL`](https://pub.dev/documentation/xl/latest/xl/XL-class.html)
##### [🐸 Zaba.app ― simple packages, simple names.](#-zabaapp--simple-packages-simple-nameshttpspubdevpublisherszabaapppackages-other-flutter-packages-published-by-zabaapp)

---

| [![XL Package Logotype](https://raw.githubusercontent.com/Zabadam/xl/master/example/android/app/src/main/res/drawable/xl.png 'Get started by adding `xl:` to your pubspec and importing `package:xl/xl.dart`')](https://pub.dev/packages/xl) | [![XL Logotype "Spinning Top"](https://raw.githubusercontent.com/Zabadam/xl/master/doc/xl.gif 'See `Logotype()` demo')](https://github.com/Zabadam/xl/blob/master/example/lib/main.dart#L39) |
| :-: | :-: |
| [![Starfield](https://raw.githubusercontent.com/Zabadam/xl/master/doc/starfield.gif 'See `ExampleStarfield()` demo')](https://github.com/Zabadam/xl/blob/master/example/lib/main.dart#L265) | [![Warpspeed](https://raw.githubusercontent.com/Zabadam/xl/master/doc/warpspeed.gif 'See `ExampleStarfield()` demo')](https://github.com/Zabadam/xl/blob/master/example/lib/main.dart#L265) |


A package providing the `XL` stack widget to implement accelerometer- and  
pointer-fueled parallax animations that smoothly transform  
children according to a spatial, three-dimensional layer definition.  
Touch and Gyroscope support included.

Design simple or elaborate *`XL`*-erometer driven interactions and interfaces.

&nbsp;

## Getting Started
To get started with sensor data, use an `XL` with `XLayer`s:
 ```dart
const Widget xl = XL(layers: [XLayer(. . .), ...]);
 ```

To distinguish between pointer data (mouse/touch) and sensors data,  
employ `PLayer`s and the relevant flags:
 ```dart
const Widget xl = XL(
  sharesPointer = false, // default is true
  sharesSensors = false, // default is false
  layers: [
    PLayer(. . .), // ignores sensors
    PLayer(. . .), // ignores sensors
    XLayer(. . .), // ignores pointer
    ...
  ],
);
 ```

&nbsp;

## Layers
The `XL.layers` property takes a list of `XLayer`s or `PLayer`s.

These `layers` allow you to define the animatable properties 
of their `child` widgets.

> [![Delicious parallax by accelerometer 🤤](https://raw.githubusercontent.com/Zabadam/xl/master/doc/example_100.gif 'Delicious parallax by accelerometer 🤤')](https://github.com/Zabadam/xl/blob/master/example/lib/main.dart#L72)
> is the result of the below `Example`:

 ```dart
class Example extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(home: Scaffold(
      body: XL(
        layers: [
          XLayer(
            xRotation: 1.0,
            yRotation: 1.0,
            xOffset: 200,
            yOffset: 200,
            child: Center(
              child: Container(
                width: 250,
                height: 250,
                color: Colors.black,
          ))),
          XLayer(
            xRotation: 1.5,
            yRotation: 1.5,
            xOffset: 300,
            yOffset: 300,
            child: Center(
              child: Container(
                width: 200,
                height: 200,
                color: Colors.red,
          ))),
        ],
      ),
    ));
  }
}
 ```

&nbsp;

## Advanced Usage
As stated, an `XL` may have both varieties of layer interspersed.  
The default function of the stack is to apply *pointer* data to both  
varieties of `Layer`, and to only apply *sensors data* to `XLayer`s;  
there are, however, flags available to toggle either functionality, as:

```dart
const layers = [
  XLayer(/* Blue X Layer; reactive */),
  PLayer(/* Red P Layer; more reactive */),
  XLayer(/* Black Layer; non-reactive control */),
];

final example = Column(children: [
  Expanded(
    child: Row(children: [
      const Expanded(
        child: XL(
          // X and P Layers behave the same way with these flags
          // (but still may differ slightly on parameters for Z axis rotation).
          sharesPointer: true,
          sharesSensors: true,
          layers: layers,
      )),
      const Expanded(
        child: XL(
          // All Layers react to pointers, but only XLayers
          // will recognize sensors. Default flags.
          sharesPointer: true, // default
          sharesSensors: false, // default
          layers: layers,
  ))])),
  Expanded(
    child: Row(children: [
      const Expanded(
        child: XL(
          // All Layers react to sensors,
          // but only PLayers will recognize pointer.
          sharesPointer: false,
          sharesSensors: true,
          layers: layers,
      )),
      const Expanded(
        child: XL(
          // Each Layer will deal only with its prescribed input data.
          sharesPointer: false,
          sharesSensors: false,
          layers: layers,
  ))])),
]);
```

> That same stack of `layers` laid out in a demonstration
> with the two flags in their four possible configurations:
>
> [![Flags to handle \`XLayers\` and \`PLayers\` separately](https://raw.githubusercontent.com/Zabadam/xl/master/doc/sharing_inputs_300.gif 'Flags to handle `X` and `P Layers` separately')](https://github.com/Zabadam/xl/blob/master/example/lib/main.dart#159)
> 
> 🟥 Red layers are `PLayer`s and react primarily to pointers data.  
> 👈 The left two 🟥 `PLayer`s also consider sensors data  
> 
> 🟦 Blue layers are `XLayer`s and react primarily to sensors data.  
> 👆 The top two 🟦 `XLayer`s also consider pointer data
> 
> ⬛ Black layers are control `XLayer`s with *no animation properties*.
> They make no reaction to any input.


&nbsp;

# Reference

## 🛣️ Roadmap
1. 🔳 Provide richer access to sensors data and ability to manipulate `Layer`s in new ways
   - Such as an `AccelSpec` and `GyroSpec` that would enable *transformation
   of* any axis based on *input from* any axis, determined by parameter.
     - Currently, for example, `yRotation` is controlled by accelerometer X axis
     rotation. `zRotation` may also be controlled by accelerometer X axis rotation
     with parameter `zRotationByX`. However, `zRotationByGyro` may interpret
     Z axis gyroscope rotations. It is the only axis with two such inputs.
     - Imagine declaring an axis to rotate or offset, then also determining
     which axis/sensor contributes to it. What deep options that could provide!
     Such as offsetting a `Widget` vertically *and* rotating it over Y axis
     based on only one input from accelerometer.
2. 🔳 Along with **#1** above, provide more Gyroscope functionality
than only one axis of influence. 
   - Gyro is a little tricker, as the sensor mostly reads flat near-`0`
values *unless the device is moving*.
   - This still provides cool options for developers to make neat interactions,
and so should be expanded upon.
3. ✅ ~~Offer an automatic `XL` stack that simply accepts a `List<Widget>`
(not necessarily `XLayer`s) and dynamically generates progressively greater
translation and rotation values for each layer.~~
   - Check out `AutoXL` which makes a parallax stack even easier to implement. \
   Simply provide any old widgets as `layers` and optionally tinker
   with the `depthFactor` or max layer `depth`.
   - Pre-configured: `AutoXL.pane`, `AutoXL.wiggler`, `AutoXL.deep`

&nbsp;

## 🐞 Bugs
1. Discrepancy with normalization delay and normalization duration.

&nbsp;

## 📜 History
*For posterity and developer preference, there are three extension "wrappers" available.*
- *Instead of the short, all-caps `XL`, consider a `ParallaxStack`*
- *Instead of the double-cap `XLayer`, consider an `AcceleraxLayer`*
- *Instead of the double-cap `PLayer`, consider a `ParallaxLayer`*

```dart
final xl = ParallaxStack( // XL
  sharesSensors: true, // same parameters and functionality
  layers: [
    AcceleraxLayer(. . .), // XLayer
    ParallaxLayer(. . .), // PLayer
  ],
);
```

##### WORK IN PROGRESS
- Furthermore, the [🌊 original website demo](https://xl.zaba.app/ 'the_wave website demo from package:mouse_parallax, converted to package:xl')
works well [with accelerometer data](https://github.com/Zabadam/xl/blob/master/website/lib/ 'the_wave website demo from package:mouse_parallax, converted to package:xl').


##### **Forked from [`mouse_parallax`](https://pub.dev/packages/mouse_parallax 'pub.dev package: mouse_parallax') considering the required `AnimatedParallaxStack` widget is not exported with the original package and that accelerometers have nothing to do with mice.**
- ##### [License here.](https://pub.dev/packages/mouse_parallax/license 'pub.dev package: mouse_parallax | license')

| ![transparent 600x1 pixel-pusher (ignore)](https://i.imgur.com/XCh0q2K.png) | [📚](#-table-of-contents 'Table of Contents') |
| --------------------------------------------------------------------------: | -------------------------------------------: |

<br />

## 🐸 [Zaba.app ― simple packages, simple names.](https://pub.dev/publishers/zaba.app/packages 'Other Flutter packages published by Zaba.app')

<details>
<summary>More by Zaba</summary>

### Wrappers | Widgets that surround other widgets with functionality
- ## 🕹️ [xl](https://pub.dev/packages/xl 'implement accelerometer-fueled interactions with a layering paradigm')
- ## 🌈 [foil](https://pub.dev/packages/foil 'implement accelerometer-reactive gradients in a cinch')
- ## 📜 [curtains](https://pub.dev/packages/curtains 'provide animated shadow decorations for a scrollable to allude to unrevealed content')

---
### Container Widget | Wraps many functionalities in one, very customizable
- ## 🌟 [surface](https://pub.dev/packages/surface 'animated, morphing container with specs for Shape, Appearance, Filter, Tactility')

---
### Succinct Utility | Work great alone or employed above
- ## 🆕 [![neu header](https://neu.zaba.app/img/neu_20.png)](https://pub.dev/packages/neu 'A helper for generating outstanding neumorphic-conforming designs')
- ## 🙋‍♂️ [img](https://pub.dev/packages/img 'An extended Image \"Too\" and DecorationImageToo that support an expanded Repeat.mirror painting mode')
- ## 🙋‍♂️ [icon](https://pub.dev/packages/icon 'An extended Icon \"Too\" for those that are not actually square, plus shadows support + IconUtils')
- ## 🏓 [ball](https://pub.dev/packages/ball 'A bouncy, position-mirroring splash factory that\'s totally customizable')
- ## 👁‍🗨 [sense](https://pub.dev/packages/sense 'A widget that outputs actionable sensors data with intervals, delays, normalization, and more.')
- ## 👥 [shadows](https://pub.dev/packages/shadows 'Convert a double-based \`elevation\` + BoxShadow and List\<BoxShadow\> extensions')
- ## 🎨 [![spectrum header](https://raw.githubusercontent.com/Zabadam/spectrum/main/doc/img/spectrum_15.gif)](https://pub.dev/packages/spectrum 'Color and Gradient utilities such as GradientTween, copyWith, complementary for colors, AnimatedGradient, MaterialColor generation, and more')
</details>


CHANGELOG.md
# 🕹️ xl
## **[0.9.1]** - 18 MAY 21
- Provide `AutoXL` to more easily generate a layered parallax stack with less setup.

## **[0.9.0]** - 13 MAY 21
- Initial release, forked from [`mouse_parallax`](https://pub.dev/packages/mouse_parallax 'pub.dev package: mouse_parallax').
  - Reimplements `ParallaxStack` with interoperability of new paradigm, but is not interchangeable with the original package.


example/lib/main.dart
import 'dart:async';
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

import 'package:xl/xl.dart';
import 'package:foil/foil.dart';

// ignore: unused_import
import 'delay.dart';
// ignore: unused_import
import 'dragging.dart';
// ignore: unused_import
import 'stack.dart';

void main() => runApp(const Example());

/// {@macro example}
class Example extends StatelessWidget {
  /// {@template example}
  /// Example parallax app utilizing the `XL` stack widget.
  /// Several demos are available.
  /// {@endtemplate}
  const Example({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) =>

      /// Comment `ExampleXL` line and uncomment another demo line ([CTRL] + [/])
      // const MaterialApp(color: Colors.black, home: Logotype());
      // const MaterialApp(color: Colors.white, home: Example0());
      // const MaterialApp(color: Colors.white, home: ExampleDragging());
      // const MaterialApp(color: Colors.white, home: ExampleStack());
      // const MaterialApp(color: Colors.white, home: ExampleDelay());
      // const MaterialApp(color: Colors.white, home: ExampleSharingInput());
      const MaterialApp(color: Colors.black, home: ExampleStarfield());
  // const MaterialApp(color: Colors.black, home: Automation());
}

/// {@macro logotype}
class Logotype extends StatelessWidget {
  /// {@template logotype}
  /// A spinning "top" made of the XL package logotype.
  /// {@endtemplate}
  const Logotype({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.black,
      body: XL(
        // useLocalPosition: false,
        layers: [
          XLayer(
            xOffset: 500,
            yOffset: 500,
            yRotation: 50,
            xRotation: 1,
            child: Center(
              child: Image(
                image: const AssetImage('res/xl.png'),
                width: 475 / (MediaQuery.of(context).devicePixelRatio / 2),
                height: 214 / (MediaQuery.of(context).devicePixelRatio / 2),
              ),
            ),
          ),
        ],
      ),
    );
  }
}

/// {@macro example0}
class Example0 extends StatelessWidget {
  /// {@template example0}
  /// A simple `XL` example with two layers.
  ///
  /// This widget reacts to pointer/touch and to sensors data.
  /// {@endtemplate}
  const Example0({Key? key}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: XL(
          layers: [
            XLayer(
                xRotation: 1.0,
                yRotation: 1.0,
                xOffset: 200,
                yOffset: 200,
                child: Center(
                    child: Container(
                  width: 250,
                  height: 250,
                  color: Colors.black,
                ))),
            XLayer(
                xRotation: 1.5,
                yRotation: 1.5,
                xOffset: 300,
                yOffset: 300,
                child: Center(
                    child: Container(
                  width: 175,
                  height: 175,
                  color: Colors.red,
                ))),
          ],
        ),
      ),
    );
  }
}

/// {@macro automation}
class Automation extends StatelessWidget {
  /// {@template automation}
  /// Generation of `XL` widgets & also presets.
  /// {@endtemplate}
  const Automation({Key? key}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    final cyans = List.generate(4, (i) => Colors.cyan[((1 + 2 * i) * 100)]);
    final purples = List.generate(4, (i) => Colors.purple[((1 + 2 * i) * 100)]);
    final pinks = List.generate(4, (i) => Colors.pink[((1 + 2 * i) * 100)]);

    List<Widget> buildLayers(int i) => [
          Container(
            width: 200.0,
            height: 200.0,
            color: purples[i],
          ),
          Container(
            width: 200.0,
            height: 200.0,
            color: pinks[i],
          ),
        ];

    final autoXLs = [
      AutoXL.pane(layers: buildLayers(0)),
      AutoXL.wiggler(layers: buildLayers(1)),
      AutoXL.deep(layers: buildLayers(2)),
      AutoXL(
        depthFactor: 100, // only accepted by default `AutoXL()`
        layers: buildLayers(3),
      ),
    ];

    return MaterialApp(
      home: Scaffold(
        backgroundColor: Colors.amber.shade100,
        body: Center(
          child: Wrap(
            children: [
              for (var i = 0; i <= 3; i++)
                Container(
                  width: 200.0,
                  height: 200.0,
                  color: cyans[i],
                  child: Center(child: autoXLs[i]),
                ),
            ],
          ),
        ),
      ),
    );
  }
}

/// {@macro example_sharing}
class ExampleSharingInput extends StatelessWidget {
  /// {@template example_sharing}
  /// Four example `XL` widgets with both an `XLayer` and
  /// `PLayer` with no autocompensation & alternating input flags.
  ///
  /// Two flags for four total configurations.
  ///
  /// - 🟥 Red layers are `PLayer`s and react primarily to pointers data.
  ///   - 👈 The left two 🟥 `PLayer`s also consider sensors data
  ///
  /// - 🟦 Blue layers are `XLayer`s and react primarily to sensors data.
  ///   - 👆 The top two 🟦 `XLayer`s also consider pointer data
  ///
  /// - ⬛ Black layers are control `XLayer`s with *no animation properties*.
  /// They make no reaction to any input.
  ///
  /// {@endtemplate}
  const ExampleSharingInput({Key? key}) : super(key: key);

  static const _layers = <PLayer>[
    XLayer(
      dimensionalOffset: 0.002,
      xRotation: 0.75,
      yRotation: 0.5,
      zRotationByX: 0.5,
      xOffset: 60,
      yOffset: 60,
      child: Center(
        child: SizedBox(
          width: 230,
          height: 350,
          child: DecoratedBox(decoration: BoxDecoration(color: Colors.blue)),
        ),
      ),
    ),
    PLayer(
      dimensionalOffset: 0.0015,
      xRotation: 1,
      yRotation: 1,
      zRotation: 1,
      xOffset: 125,
      yOffset: 125,
      child: Center(
        child: SizedBox(
          width: 150,
          height: 200,
          child: DecoratedBox(
            decoration: BoxDecoration(
              color: Colors.red,
              boxShadow: [BoxShadow(blurRadius: 20, spreadRadius: -6)],
            ),
          ),
        ),
      ),
    ),
    XLayer(
      child: Center(
        child: SizedBox(
          width: 75,
          height: 75,
          child: DecoratedBox(decoration: BoxDecoration(color: Colors.black)),
        ),
      ),
    )
  ];

  @override
  Widget build(BuildContext context) => Scaffold(
        body: Column(
          children: [
            Expanded(
                child: Row(children: [
              const Expanded(
                  child: XL(
                sharesPointer: true,
                sharesSensors: true,
                layers: _layers,
              )),
              const Expanded(
                  child: XL(
                sharesPointer: true, // default
                sharesSensors: false, // default
                layers: _layers,
              ))
            ])),
            Expanded(
                child: Row(children: [
              const Expanded(
                  child: XL(
                sharesPointer: false,
                sharesSensors: true,
                layers: _layers,
              )),
              const Expanded(
                  child: XL(
                sharesPointer: false,
                sharesSensors: false,
                layers: _layers,
              ))
            ])),
          ],
        ),
      );
}

/// {@macro example_starfield}
class ExampleStarfield extends StatefulWidget {
  /// {@template example_starfield}
  /// 🎊🚀🌌 Warp Speed!
  ///
  /// Zoom and pan this InteractiveViewer containing a generated "starfield",
  /// replete with randomness and parallax.
  ///
  /// Parallax for this demo is based on accelerometer only.
  /// Pointer is supported for the InteractiveViewer only.
  /// The FAB will `setState()` and generate a new starfield.
  /// {@endtemplate}
  const ExampleStarfield({Key? key}) : super(key: key);

  @override
  _ExampleStarfieldState createState() => _ExampleStarfieldState();
}

class _ExampleStarfieldState extends State<ExampleStarfield> {
  @override
  Widget build(BuildContext context) {
    const _duration = Duration(milliseconds: 250);
    const _ms = Duration(milliseconds: 1);
    final random = Random();
    final _starfield = <XL>[];

    final _colors = List.from(Colors.primaries)..shuffle();
    final _alignments = [
      Alignment.bottomCenter,
      Alignment.bottomLeft,
      Alignment.bottomRight,
      Alignment.center,
      Alignment.centerLeft,
      Alignment.centerRight,
      Alignment.topCenter,
      Alignment.topLeft,
      Alignment.topRight,
      Alignment(random.nextBool() ? 1 : -1 * random.nextDouble() * 6.0,
          random.nextBool() ? 1 : -1 * random.nextDouble() * 6.0),
      Alignment(random.nextBool() ? 1 : -1 * random.nextDouble() * 6.0,
          random.nextBool() ? 1 : -1 * random.nextDouble() * 6.0),
      Alignment(random.nextBool() ? 1 : -1 * random.nextDouble() * 6.0,
          random.nextBool() ? 1 : -1 * random.nextDouble() * 6.0),
    ]..shuffle();
    final _gradients = [
      Foils.gymClassParachute,
      Foils.linearLooping,
      Foils.linearLoopingReversed,
      Foils.linearRainbow,
      Foils.linearReversed,
      Foils.oilslick,
      Foils.rainbow,
      Foils.sitAndSpin,
      Foils.stepBowLinear,
      Foils.stepBowRadial,
      Foils.stepBowSweep
    ]..shuffle();

    for (var xl = 0; xl < 20; xl++) {
      final duration = _duration + _ms * random.nextInt(2500);
      final d = random.nextDouble();

      _starfield.add(
        XL(
          // This example is in an InteractiveViewer with its own
          // gesture detection, so this is an accelerometer-only demo.
          sharesPointer: false,
          sharesSensors: true,
          duration: duration,
          dragging: Dragging(duration: duration * 2), // FIXME ✝
          normalization:
              const Normalization(delay: Duration(milliseconds: 2500)),
          layers: [
            for (var layer = 0; layer < 25; layer++)
              XLayer(
                dimensionalOffset: 0.05,
                xRotation: 0.5,
                yRotation: 0.5,
                zRotationByX: 20,
                xOffset: 0,
                yOffset: 0,
                child: AnimatedAlign(
                  alignment: (_alignments..shuffle())[0] +
                      Alignment(random.nextDouble(), random.nextDouble()) -
                      Alignment(random.nextDouble(), random.nextDouble()),
                  duration: duration,
                  child: Foil.sheet(
                    opacity: 0.5,
                    duration: duration,
                    speed: const Duration(milliseconds: 250),
                    gradient: (_gradients..shuffle())[0],
                    sheet: Sheet(
                      width: (xl + 1) * d + layer / 25 - layer / 25 * d,
                      height: (xl + 1) * d + layer / 25 - layer / 25 * d,
                      decoration: BoxDecoration(
                        shape: BoxShape.circle,
                        color: (_colors..shuffle())[0],
                      ),
                    ),
                  ),
                ),
              )
          ],
        ),
      );
    }

    SystemChrome.setSystemUIOverlayStyle(
      const SystemUiOverlayStyle(
        statusBarColor: Colors.transparent,
        systemNavigationBarColor: Colors.transparent,
        systemNavigationBarDividerColor: Colors.transparent,
        systemNavigationBarIconBrightness: Brightness.light,
      ),
    );

    return Scaffold(
      extendBody: true,
      resizeToAvoidBottomInset: false,
      backgroundColor: Colors.black,
      floatingActionButton:
          FloatingActionButton(onPressed: () => setState(() {})),
      body: InteractiveViewer(
        boundaryMargin: const EdgeInsets.all(double.infinity),
        minScale: 0.01,
        maxScale: 12.5,
        child: Roll(
          crinkle: Crinkle.twinkling,
          child: Stack(children: [for (final xl in _starfield) xl]),
        ),
      ),
    );
  }
}

/// * # ✝
///
/// # FIXME
///
/// Despite not using pointer data to influence the parallax effect of
/// `XLayer`s by [XL.sharesPointer] set `false`,
/// a hovering pointer (or touched finger) will still influence the
/// parallax animation duration.
///
/// Work around this here by setting the [Dragging.duration]
/// to double that of the [XL.duration].
///
///
/// See `src/xl.dart` for current issue with normalization duration:
///
/// ```
/// SensorListener(
///   ...
///   normalizationDuration: widget.duration * 10, // FIXME
///   ...
/// ```
///
/// ## AND
///
/// ```
/// AnimatedXL(
///   ...
///   duration: hovering
///     ? widget.dragging.duration
///     : normalizing
///         ? widget.duration * 2 // FIXME
///         : widget.duration,
///   ...
/// ```
const markdownHover = null;
